papaya-web/src/main/java/pl/edu/agh/papaya/rest/projects/controller/ProjectsController.java

```diff
@@ -90,12 +95,75 @@ private Project getValidProject(Long id) {
                 .stream()
                 .sorted(Comparator.comparing(sprint -> sprint.getEnrollmentPeriod().getStart()))
                 .map(sprint -> sprintMapper.mapToApi(sprint, currentTime))
                 .collect(Collectors.toList());
 
         return ResponseEntity.ok(sprints);
     }
+
+    @Override
+    public ResponseEntity<AvailabilityDto> getUserAvailability(Long projectId, Long sprintId) {
+        Project project = getValidProject(projectId);
+        Sprint sprint = getValidSprint(sprintId);
+
+        if (!isSprintInProject(sprint, project)) {
+            throw new ForbiddenAccessException();
+        }
+
+        Availability availability = availabilityService
+                .getBySprintIdAndUserId(sprintId, userContext.getUserIdn())
+                .orElseThrow(ResourceNotFoundException::new);
+
+        return ResponseEntity.ok(availabilityMapper.mapToApi(availability));
+    }
+
+    @Override
+    public ResponseEntity<AvailabilityDto> updateUserAvailability(@Valid AvailabilityDto body, Long projectId,
+            Long sprintId) {
+        Project project = getValidProject(projectId);
+        Sprint sprint = getValidSprint(sprintId);
+        if (!isSprintInProject(sprint, project) || !isAvailabilityDeclarationAllowed(sprint)
+                || !body.getUserId().equals(userContext.getUserIdn())) {
+            throw new ForbiddenAccessException();
+        }
+
+        Availability availability = availabilityService.getBySprintIdAndUserId(sprintId, userContext.getUserIdn())
+                .map(persistedAvailability -> updatePersistedAvailability(persistedAvailability, body))
+                .orElseGet(() -> createAvailability(body, sprint));
+
+        return ResponseEntity.ok(availabilityMapper.mapToApi(availability));
+    }
+
+    private Sprint getValidSprint(Long sprintId) {
+        return sprintService.getById(sprintId)
+                .orElseThrow(ResourceNotFoundException::new);
+    }
+
+    private boolean isSprintInProject(Sprint sprint, Project project) {
+        return sprint.getProject().getId().equals(project.getId());
+    }
+
+    private boolean isAvailabilityDeclarationAllowed(Sprint sprint) {
+        return sprint.getSprintState() == SprintState.DECLARABLE;
+    }
+
+    private Availability createAvailability(AvailabilityDto body, Sprint sprint) {
+        return availabilityService.newAvailability()
+                .withSprint(sprint)
+                .withUser(userContext.getUser())
+                .withTimeAvailable(Duration.ofMinutes(body.getTimeAvailable()))
+                .withPastSprintRemainingTime(Duration.ofMinutes(body.getTimeRemaining()))
+                .withNotes(body.getNotes())
+                .create();
+    }
+
+    private Availability updatePersistedAvailability(Availability availability, AvailabilityDto body) {
+        availability.setNotes(body.getNotes());
```

Comment:
Upewniłabym się, że `notes` nie jest nullem

Reply:
nic nie stoi na przeszkodzie, żeby notes było nullem

Reply:
No właśnie wiem - chcemy nadpisać nullem, jak przyjdzie taki PUT?

Reply:
to by chyba znaczyło, że albo były nullem i zostały nullem, albo że zostały usunięte

Reply:
wydaje mi się, że to nie problem

Reply:
albo notes zrobić required... tylko że wtedy chyba trzeba do mappera dodać zwracanie pustego stringa w przypadku nulla?

Reply:
JUŻ NIGDY